// @automattic/vip/dist/bin/vip-config-envvar-get-all.js
#!/usr/bin/env node
"use strict";

exports.__esModule = true;
exports.getAllEnvVarsCommand = getAllEnvVarsCommand;
var _chalk = _interopRequireDefault(require("chalk"));
var _command = _interopRequireDefault(require("../lib/cli/command"));
var _format = require("../lib/cli/format");
var _api = require("../lib/envvar/api");
var _logging = require("../lib/envvar/logging");
var _tracker = require("../lib/tracker");
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const exampleUsage = 'vip @example-app.develop config envvar get-all';
const usage = 'vip config envvar get-all';

// Command examples
const examples = [{
  usage: exampleUsage,
  description: 'Retrieve a list of all environment variables in the default table format.'
}, {
  usage: `${exampleUsage} --format=csv`,
  description: 'Retrieve a list of all environment variables in CSV format.'
}, {
  usage: `${exampleUsage} --format=ids`,
  description: 'Retrieve a list of all environment variable names as a space separated list.'
}, {
  usage: `${exampleUsage} --format=keyValue`,
  description: 'Retrieve a list of all environment variables as a key value list.'
}];

/**
 * @param {string[]} arg
 * @param {object} opt
 * @return {Promise<void>}
 */
async function getAllEnvVarsCommand(arg, opt) {
  const trackingParams = {
    app_id: opt.app.id,
    command: usage,
    env_id: opt.env.id,
    format: opt.format,
    org_id: opt.app.organization.id
  };
  (0, _logging.debug)(`Request: Get all environment variables for ${(0, _logging.getEnvContext)(opt.app, opt.env)}`);
  await (0, _tracker.trackEvent)('envvar_get_all_command_execute', trackingParams);
  const envvars = await (0, _api.getEnvVars)(opt.app.id, opt.env.id).catch(async err => {
    await (0, _tracker.trackEvent)('envvar_get_all_query_error', {
      ...trackingParams,
      error: err.message
    });
    throw err;
  });
  await (0, _tracker.trackEvent)('envvar_get_all_command_success', trackingParams);
  if (0 === envvars.length) {
    console.log(_chalk.default.yellow('There are no environment variables'));
    process.exit();
  }

  // Vary data by expected format.
  let key = 'name';
  if ('keyValue' === opt.format) {
    key = 'key';
  } else if ('ids' === opt.format) {
    key = 'id';
  }
  const envvarsObject = envvars.map(({
    name: envvarName,
    value
  }) => ({
    [key]: envvarName,
    value
  }));
  console.log((0, _format.formatData)(envvarsObject, opt.format));
}
(0, _command.default)({
  appContext: true,
  appQuery: _api.appQuery,
  envContext: true,
  format: true,
  usage
}).examples(examples).argv(process.argv, getAllEnvVarsCommand);
