// @automattic/vip/dist/lib/app-logs/app-logs.js
"use strict";

exports.__esModule = true;
exports.LIMIT_MAX = void 0;
exports.getRecentLogs = getRecentLogs;
var _graphqlTag = _interopRequireDefault(require("graphql-tag"));
var _api = _interopRequireDefault(require("../../lib/api"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const LIMIT_MAX = exports.LIMIT_MAX = 5000;
const QUERY_ENVIRONMENT_LOGS = (0, _graphqlTag.default)`
	query GetAppLogs(
		$appId: Int
		$envId: Int
		$type: AppEnvironmentLogType
		$limit: Int
		$after: String
	) {
		app(id: $appId) {
			environments(id: $envId) {
				id
				logs(type: $type, limit: $limit, after: $after) {
					nodes {
						timestamp
						message
					}
					nextCursor
					pollingDelaySeconds
				}
			}
		}
	}
`;
async function getRecentLogs(appId, envId, type, limit, after) {
  const api = (0, _api.default)({
    exitOnError: false
  });
  const response = await api.query({
    query: QUERY_ENVIRONMENT_LOGS,
    variables: {
      appId,
      envId,
      type,
      limit,
      after
    }
  });
  const logs = response.data.app?.environments?.[0]?.logs;
  if (!logs?.nodes) {
    throw new Error('Unable to query logs');
  }
  return logs;
}
