// @automattic/vip/dist/lib/api/app.js
"use strict";

exports.__esModule = true;
exports.default = _default;
var _graphqlTag = _interopRequireDefault(require("graphql-tag"));
var _api = _interopRequireDefault(require("../../lib/api"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
async function _default(app, fields = 'id,name', fragments = '') {
  const api = (0, _api.default)();
  if (isNaN(Number(app))) {
    const res = await api.query({
      query: (0, _graphqlTag.default)`query App( $name: String ) {
					apps( first: 1, name: $name ) {
						total,
						nextCursor,
						edges {
							${fields}
						}
					}
				}
				${fragments || ''}`,
      variables: {
        name: app
      }
    });
    if (!res.data.apps?.edges?.length) {
      return {};
    }
    return res.data.apps.edges[0];
  }
  if (typeof app === 'string') {
    app = parseInt(app, 10);
  }
  const appQuery = {
    query: (0, _graphqlTag.default)`query App( $id: Int ) {
				app( id: $id ){
					${fields}
				}
			}
			${fragments || ''}`,
    variables: {
      id: app
    }
  };
  const customDeployToken = process.env.WPVIP_DEPLOY_TOKEN;
  if (customDeployToken) {
    appQuery.context = {
      headers: {
        Authorization: `Bearer ${customDeployToken}`
      }
    };
  }
  const res = await api.query(appQuery);
  if (!res.data.app) {
    return {};
  }
  return res.data.app;
}
