// @automattic/vip/dist/lib/analytics/index.js
"use strict";

exports.__esModule = true;
exports.default = void 0;
var _debug = _interopRequireDefault(require("debug"));
var _env = _interopRequireDefault(require("../env"));
function _interopRequireDefault(e) { return e && e.__esModule ? e : { default: e }; }
const debug = (0, _debug.default)('@automattic/vip:analytics');

/* eslint-disable camelcase */
const client_info = {
  cli_version: _env.default.app.version,
  os_name: _env.default.os.name,
  os_version: _env.default.os.version,
  node_version: _env.default.node.version
};
/* eslint-enable camelcase */

class Analytics {
  clients;
  constructor(clients) {
    this.clients = clients;
  }
  async trackEvent(name, props = {}) {
    if (process.env.DO_NOT_TRACK) {
      debug(`trackEvent() for ${name} => skipping per DO_NOT_TRACK variable`);
      return [];
    }
    return Promise.all(this.clients.map(client => client.trackEvent(name, {
      // eslint-disable-next-line camelcase
      ...client_info,
      ...props
    })));
  }
}
exports.default = Analytics;
